File: src/test/java/cz/vhromada/catalog/validator/PictureValidatorTest.java
package cz.vhromada.catalog.validator;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.SoftAssertions.assertSoftly;
import static org.mockito.Mockito.verifyZeroInteractions;

import java.util.Collections;

import cz.vhromada.catalog.entity.Picture;
import cz.vhromada.catalog.utils.PictureUtils;
import cz.vhromada.common.Movable;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.validator.MovableValidatorTest;
import cz.vhromada.common.validator.MovableValidator;
import cz.vhromada.common.validator.ValidationType;
import cz.vhromada.result.Event;
import cz.vhromada.result.Result;
import cz.vhromada.result.Severity;
import cz.vhromada.result.Status;

import org.junit.jupiter.api.Test;

/**
 * A class represents test for class {@link PictureValidator}.
 *
 * @author Vladimir Hromada
 */
class PictureValidatorTest extends MovableValidatorTest<Picture, cz.vhromada.catalog.domain.Picture> {

    /**
     * Test method for {@link PictureValidator#PictureValidator(MovableService)} with null service for pictures.
     */
    @Test
    void constructor_NullPictureService() {
        assertThatThrownBy(() -> new PictureValidator(null)).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link PictureValidator#validate(Movable, ValidationType...)} with {@link ValidationType#DEEP} with data with null name.
     */
    @Test
    void validate_Deep_NullContent() {
        final Picture picture = getValidatingData(1);
        picture.setContent(null);

        final Result<Void> result = getMovableValidator().validate(picture, ValidationType.DEEP);

        assertSoftly(softly -> {
            softly.assertThat(result.getStatus()).isEqualTo(Status.ERROR);
            softly.assertThat(result.getEvents())
                .isEqualTo(Collections.singletonList(new Event(Severity.ERROR, "PICTURE_CONTENT_NULL", "Content mustn't be null.")));
        });

        verifyZeroInteractions(getMovableService());
    }

    @Override
    protected MovableValidator<Picture> getMovableValidator() {
        return new PictureValidator(getMovableService());
    }

    @Override
    protected Picture getValidatingData(final Integer id) {
        return PictureUtils.newPicture(id);
    }

    @Override
    protected Picture getValidatingData(final Integer id, final Integer position) {
        final Picture picture = PictureUtils.newPicture(id);
        picture.setPosition(position);

        return picture;
    }

    @Override
    protected cz.vhromada.catalog.domain.Picture getRepositoryData(final Picture validatingData) {
        return PictureUtils.newPictureDomain(validatingData.getId());
    }

    @Override
    protected cz.vhromada.catalog.domain.Picture getItem1() {
        return PictureUtils.newPictureDomain(1);
    }

    @Override
    protected cz.vhromada.catalog.domain.Picture getItem2() {
        return PictureUtils.newPictureDomain(2);
    }

    @Override
    protected String getName() {
        return "Picture";
    }

}

File: src/test/kotlin/cz/vhromada/catalog/validator/PictureValidatorTest.kt
package cz.vhromada.catalog.validator

import com.nhaarman.mockitokotlin2.verifyZeroInteractions
import cz.vhromada.catalog.entity.Picture
import cz.vhromada.catalog.utils.PictureUtils
import cz.vhromada.common.result.Event
import cz.vhromada.common.result.Severity
import cz.vhromada.common.result.Status
import cz.vhromada.common.test.validator.MovableValidatorTest
import cz.vhromada.common.validator.AbstractMovableValidator
import cz.vhromada.common.validator.MovableValidator
import cz.vhromada.common.validator.ValidationType
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.junit.jupiter.api.Test

/**
 * A class represents test for class [PictureValidator].
 *
 * @author Vladimir Hromada
 */
class PictureValidatorTest : MovableValidatorTest<Picture, cz.vhromada.catalog.domain.Picture>() {

    /**
     * Test method for [AbstractMovableValidator.validate] with [ValidationType.DEEP] with data with null content.
     */
    @Test
    fun validateDeepNullContent() {
        val picture = getValidatingData(1)
                .copy(content = null)

        val result = getValidator().validate(picture, ValidationType.DEEP)

        assertSoftly {
            it.assertThat(result.status).isEqualTo(Status.ERROR)
            it.assertThat(result.events()).isEqualTo(listOf(Event(Severity.ERROR, "PICTURE_CONTENT_NULL", "Content mustn't be null.")))
        }

        verifyZeroInteractions(service)
    }

    override fun getValidator(): MovableValidator<Picture> {
        return PictureValidator(service)
    }

    override fun getValidatingData(id: Int?): Picture {
        return PictureUtils.newPicture(id)
    }

    override fun getValidatingData(id: Int?, position: Int?): Picture {
        return PictureUtils.newPicture(id)
                .copy(position = position)
    }

    override fun getRepositoryData(validatingData: Picture): cz.vhromada.catalog.domain.Picture {
        return PictureUtils.newPictureDomain(validatingData.id)
    }

    override fun getItem1(): cz.vhromada.catalog.domain.Picture {
        return PictureUtils.newPictureDomain(1)
    }

    override fun getItem2(): cz.vhromada.catalog.domain.Picture {
        return PictureUtils.newPictureDomain(2)
    }

    override fun getName(): String {
        return "Picture"
    }

}
